AST-verify that 'open' is only used where appropriate

jrose-apple opened the pull request:
...and fix places where it was being used inappropriately. This should have no effect in practice, but it does change the formal access level of some declarations from 'open' to 'public' (where 'open' didn't make sense) and from 'internal' down to 'fileprivate' (where 'internal' was overly generous for some synthesized declarations).

jrose-apple commented:
Hm, this will probably break LLDB…

slavapestov commented on include/swift/AST/Decl.h:
```diff
@@ -6703,7 +6703,9 @@ inline bool ValueDecl::isImportAsMember() const {
       isa<SubscriptDecl>(this) ||
-      isa<FuncDecl>(this)) {
+      isa<FuncDecl>(this) ||
+      isa<AccessorDecl>(this) ||
```
AccessorDecl is a FuncDecl

jrose-apple replied:
Oops, I thought it was just an AbstractFunctionDecl. Thanks.

gottesmm commented:
@jrose-apple Can you elaborate in more detail on what "appropriate means". That seems like a vey ambiguous phrase.

slavapestov commented on lib/AST/ASTVerifier.cpp:
```diff
@@ -2114,11 +2114,20 @@ class Verifier : public ASTWalker {
-        Out << " does not have access";
-        abort();
+      if (VD->hasAccess()) {
+        if (VD->getFormalAccess() == AccessLevel::Open &&
```
Could add a similar check for 'final'

jrose-apple replied:
I'm scared that'll turn up more places where it's wrong, but I guess that's the point.

slavapestov reviewed (commented):
Looks great otherwise.

jrose-apple commented on lib/AST/Decl.cpp:
```diff
@@ -2282,7 +2282,8 @@ void ValueDecl::copyFormalAccessFrom(const ValueDecl *source,
       !getAttrs().hasAttribute<UsableFromInlineAttr>() &&
-      !getAttrs().hasAttribute<InlinableAttr>()) {
+      !getAttrs().hasAttribute<InlinableAttr>() &&
+      DeclAttribute::canAttributeAppearOnDecl(DAK_UsableFromInline, this)) {
```
Sure, I can add one. The particular case was a synthesized typealias, which shows up in the stdlib.

jrose-apple commented:
> Can you elaborate in more detail on what "appropriate means".

That's in the code. I don't think it really needs to be in the commit message.

gottesmm commented:
@jrose-apple Can you elaborate on why you think that is sufficient?

My reasoning for pushing back on this is that IMO not expressing your intent very explicitly in a commit message leads to siloing of information since language design (and language designer intentions) become intertwined with the code base. Thus if one does not have a high level overview of the code base or spent much time with this part of the codebase, the semantic intention is lost.

jrose-apple commented:
I…don't disagree with your high-level comment, but neither do I think repeating the content of the commit in the commit message would help here, and I don't have anything *higher-*level than the content of the commit. It really is just "classes and potentially-overridable decls". There's no unifying rule besides that.

gottesmm commented:
Then I would say in the commit message:

1. Specifically that you are fixing this for "classes and potentially-overrideable decls".
2. What were the old semantics that were incorrect... how are they incorrect and why were these incorrect semantics introduced in the first place?
3. What are the new semantics, why are they correct.

The reason why it is useful to put this information in the commit message is that:

1. The possibility of your intent being misunderstood has been eliminated.
2. It also helps to eliminate the problem of "what is actually 'correct swift'... what were the actual 'intended semantics'". If the answer is what Jordan can infer from this commit that he made 5 years ago, I think we have a problem. As you know there have been more than a couple of situations where due to forgetfulness, the actual expected semantics have been forgotten by engineers.

jrose-apple commented:
There isn't any semantics change here. There's an implicit rule about what `open` can be applied to, based on the SE proposal that introduced it. I added a verifier check that enforces this rule, and then fixed the handful of places that were violating it. If the rule is wrong, that's an SE issue; if the "fixes" are wrong, that's something that needs to be rederived for each use.

slavapestov commented:
I agree that perhaps saying “verify that ‘open’ only appears on overridable declarations in classes” is better than “where appropriate”, but I don’t expect every commit message to summarize the relevant part of the language design. This is not like some of the more esoteric corner cases we deal with - the meaning of “open” was debated in great detail on the evolution list and is well understood by developers and users.

jrose-apple committed:
AST-verify that 'open' is only used on classes and overridable members

...and fix places where it was being used inappropriately.

- Don't use 'open' on non-class members in the importer.
- Use the existing 'copyFormalAccessFrom' instead of an ad hoc version
  for synthesized typealiases for protocol conformances. (This can
  change 'internal' down to 'fileprivate', but only where the
  enclosing type was already 'private' or 'fileprivate'.)
- Fix 'copyFormalAccessFrom' to not copy '@usableFromInline' onto
  declarations that don't support it (namely, the above typealiases).

This should have no visible effect in practice.

jrose-apple commented:
Okay, reworded the commit message and fixed the AccessorDecl thing. Still to do: get LLDB to follow these rules, also write a test that shows it isn't following these rules.

The `final` thing uncovered the fact that we weren't rejecting `open let`, so I'll do that in a follow-up PR instead.

jrose-apple commented:
@swift-ci Please smoke test

gottesmm commented:
@jrose-apple Great comment! = D.

gottesmm commented:
erm commit msg ; )
